modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java

```diff
@@ -0,0 +1,150 @@
+        this.valuesSources = valuesSources;
+        if (valuesSources != null && !valuesSources.isEmpty()) {
+            stats = context.bigArrays().newObjectArray(1);
+        }
+    }
+
+    @Override
+    public boolean needsScores() {
+        boolean needsScores = false;
+        if (valuesSources != null) {
+            for (Map.Entry<String, ValuesSource.Numeric> valueSource : valuesSources.entrySet()) {
+                needsScores |= valueSource.getValue().needsScores();
+            }
+        }
+        return needsScores;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                final LeafBucketCollector sub) throws IOException {
+        if (valuesSources == null || valuesSources.isEmpty()) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final BigArrays bigArrays = context.bigArrays();
+        final HashMap<String, SortedNumericDoubleValues> values = new HashMap<>(valuesSources.size());
+        for (Map.Entry<String, ValuesSource.Numeric> valuesSource : valuesSources.entrySet()) {
+            values.put(valuesSource.getKey(), valuesSource.getValue().doubleValues(ctx));
+        }
+
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                // get fields
+                Map<String, Double> fields = getFields(doc);
+                if (fields != null) {
+                    stats = bigArrays.grow(stats, bucket + 1);
+                    RunningStats stat = stats.get(bucket);
+                    // add document fields to correlation stats
+                    if (stat == null) {
+                        stat = new RunningStats(fields);
+                    } else {
+                        stat.add(fields);
+                    }
+                    stats.set(bucket, stat);
+                }
+            }
+
+            /**
+             * return a map of field names and data
+             */
+            private Map<String, Double> getFields(int doc) {
+                // get fieldNames to use as hash keys
+                ArrayList<String> fieldNames = new ArrayList<>(values.keySet());
+                HashMap<String, Double> fields = new HashMap<>(fieldNames.size());
+
+                // loop over fields
+                for (String fieldName : fieldNames) {
+                    final SortedNumericDoubleValues doubleValues = values.get(fieldName);
+                    doubleValues.setDocument(doc);
+                    final int valuesCount = doubleValues.count();
+                    // if document contains an empty field we omit the doc from the correlation
+                    if (valuesCount <= 0) {
+                        return null;
+                    }
+                    // get the field value (multi-value is the average of all the values)
```

Comment:
I think thorowing an exception would be a good first step and then in a follow-up we could consider accepting a `mode` option like sorting does so that users can configure that we should take the min/max/median/average in case of multiple values